### ISSUE TITLE
Dragging a file from Finder into the Terminal results in unescaped file path

### ISSUE THREAD

**Issue by krackers** [fish-shell/fish-shell#4323]
fish, version 2.6.0
Using Apple Terminal 2.5 on 10.9.5

When you drag a file from the finder into the Terminal, on bash you get the file path escaped properly, especially in spaces: for instance `/Users/blah/Downloads/folder\ with\ spaces`. However, on fish you get the path literally without any spaces escaped.

This is probably related to the bracketed paste feature, since a similar issue [afflicts](https://apple.stackexchange.com/questions/223948/dragging-and-dropping-a-file-onto-a-terminal-window-doesn-t-escape-the-path-unde) zsh. Apparently zsh has a feature called `bracketed-paste-magic` that can solve this, but I haven't tried it. 

I saw these other discussions about pasting https://github.com/fish-shell/fish-shell/issues/967 https://github.com/fish-shell/fish-shell/issues/2188 that might be related and discuss that bracketed paste magic (though I'm not sure if dragging files onto the terminal is implemented differently from pasting). 

Also interestingly the answer from the stackexchange link about holding control while letting go of the mouse after dragging seems to force shell quoting

**Comment by faho** [comment]
>(though I'm not sure if dragging files onto the terminal is implemented differently from pasting).

Most likely not.

We have conciously stayed away from escaping all pastes, because people often paste actual shell code, and you do not want that escaped.

What we have instead, is the following:

When the cursor is inside single-quotes (including if there is an unclosed single-quote), any paste will have both single-quotes (`'`) and backslashes (`\`) escaped, so you can simply close the quote to use it in a single token.

So try entering `'`, then dragging the file, and then another `'`.

If that doesn't work, this requires some special support. Which I would be open to implement (though I can't test it) once someone points me to some documentation.

**Comment by krader1961** [comment]
Works for me. By which I mean that the path has the spaces escaped by backslashes. I'm using macOS 10.12 and fish built from git head. Here I typed `echo ` at my fish prompt then dragged a long file name containing spaces from Finder to the terminal:

```
15:23 macbook ip36  ~/p/3/fish-shell  (var_entry_t)
echo /Users/krader/Music/Artists+Albums/Lake\ Street\ Dive/Bad\ Self\ Portraits/01\ -\ Bad\ Self\ Portraits.mp3
```

OS X 10.9 is four years old and it looks like newer versions do the right thing. Any chance you can upgrade your OS? This [Superuser](https://superuser.com/questions/829295/how-to-copy-unix-file-path-in-os-x-yosemite) question implies this was fixed in OS X three years ago so it may be sufficient to upgrade to OS X 10.10.

**Comment by krackers** [comment]
@fasho
No entering it in quotes doesn't work either, although the issue is kind of moot at that point since you may as if you wrap it in double quotes it doesn't matter if it's escaped or not.

@krader1961 
I need to stay on 10.9 at the moment. Is there a way to disable bracketed paste at least since this issue is really annoying.

**Comment by krader1961** [comment]
I'm pretty sure this has nothing to do with bracketed paste. But you can disable it by defining your own functions that enable/disable it. There isn't an easy way to disable it but you can do so by editing `$__fish_datadir/__fish_config_interactive.fish`. Find the block beginning with `if not set -q FISH_UNIT_TESTS_RUNNING`. You should see that it defines two functions named `__fish_enable_bracketed_paste` and `__fish_disable_bracketed_paste`. Replace those with empty functions. That is,

```
function __fish_enable_bracketed_paste; end
function __fish_disable_bracketed_paste; end
```

Let us know if that changes the behavior on your system since it shouldn't.

**Comment by krackers** [comment]
@krader1961 
Yes it does indeed change the behavior -- the paths are escaped correctly now.

**Comment by faho** [comment]
I'm closing this since it's a bug in Terminal.app and a workaround has been provided (either that, upgrade your OS or switch to a different terminal).

I don't think there is anything we should do about it.

**Comment by krackers** [comment]
Yea you're right. I took a look into the disassembled Terminal.app and saw this key method that handles the logic 

`TTView readFilePathFromString:(NSString*)path escapeTilde:(BOOL)shouldEscape isCommand:(BOOL)commandQ forceEscape:(BOOL)shouldForce`

In there it was set up so that if bracketed paste was enabled then the file path would not be escaped. This behavior was fixed in subsequent versions. To simulate the same thing one can either modify the assembly directly (risky) or create a SIMBL plugin that swizzles the relevant method. I was already using MouseTermPlus so it was trivial to just swizzle that method as well, setting shouldForce to always be true.

```
- (BOOL) MouseTerm_readFilePathFromString:(NSString*)path escapeTilde:(BOOL)shouldEscape isCommand:(BOOL)commandq forceEscape:(BOOL)shouldForce
{
    return [self MouseTerm_readFilePathFromString:path escapeTilde:shouldEscape isCommand:commandq forceEscape:TRUE];
}
```
